### PR TITLE
fix for #7252

### PR DESCRIPTION
**Pull request by krux02**
```nim
let v32:float32 = 1.00001
let v64:float64 = 1.0000001
echo v32
echo v64
write(stdout, v32)
echo ""
write(stdout, v64)
```

output:
```
1.000010013580322
1.0000001
1.000010013580322
1.0000001
```

At least it is consistent now.

### CODE REVIEW

**Comment by Araq** [comment]
- Do not use semicolons.
- I don't like your rewrite. The easier fix is to use "%.16g" in the write procs.

**Comment by krux02** [comment]
What is the problem with the rewrite. I made sure that all float to string/stream/file functions use the same code for formating.

**Comment by krux02** [comment]
I removed the semicolon, but to be honest I don't 100% understand the problem with nimscript. I don't use nimscript and the test fails with the following problem:

```
$ nim e tests/test_nimscript.nims
Hint: used config file '/local/doering/Nim/config/nim.cfg' [Conf]
Hint: used config file '/local/doering/Nim/tests/test_nimscript.nims' [Conf]
lib/system/sysio.nim(205, 11) Error: undeclared identifier: 'writeFloatToBuffer'
```

File: lib/system/sysstr.nim [hunk 1]
```diff
@@ -325,41 +325,46 @@ proc nimIntToStr(x: int): string {.compilerRtl.} =
...
+  var n: int = c_sprintf(buf[0].addr, "%.16g", value)
+  var hasDot = false
+  for i in 0..n-1:
+    if buf[i] == ',':
+      buf[i] = '.'
+      hasDot = true
+    elif buf[i] in {'a'..'z', 'A'..'Z', '.'}:
+      hasDot = true
+  if not hasDot:
+    return n + c_sprintf(buf[n].addr, ".0")
```
Comment (Araq) [inline lib/system/sysstr.nim]: This and the following ``c_sprintf`` are to be replaced by proper append operations.

Comment (krux02) [reply lib/system/sysstr.nim]: how should I do that? Which append proc are you referring to? It's not an appendable array. It is a static array. I could write the following: 

```nim    
  buf[n .. n+2] = ".0\0"
  return n + 2
```
But that operator isn't even supported, nor do I think it's better readable. With the sprintf and the index n in the argument I think I make it pretty clear that I want to append the string including the 0-terminator.

File: changelog.md [hunk 2]
```diff
@@ -33,9 +33,9 @@
   implemented. These are now implemented purely in ``system.nim`` without compiler
   support. There is a new "heterogenous" slice type ``system.HSlice`` that takes 2
   generic parameters which can be ``BackwardsIndex`` indices. ``BackwardsIndex`` is
-  produced by ``system.^``.
-  This means if you overload ``[]`` or ``[]=`` you need to ensure they also work
-  with ``system.BackwardsIndex`` (if applicable for the accessors).
+  produced by ``system.^``.  Using the ``BackwardsIndex`` on arrays
```
Comment (Araq) [inline changelog.md]: This is unrelated to your PR.

Comment (krux02) [reply changelog.md]: yes you are right. it was supposed to be in the other PR. Oops

**Comment by Araq** [comment]
Fixed it differently.
